### ISSUE TITLE
CoreText renderer doesn't do font fallback

### ISSUE THREAD

**Issue by kballard** [macvim-dev/macvim#243]
When using the CoreText renderer, MacVim doesn't appear to do any font fallback. Glyphs that don't exist in the current font are always rendered with the missing glyph icon instead of invoking system font fallback to find some font that can supply the glyph.

This used to work. But I'm not quite sure when it broke. Either MacVim used to do font fallback and stopped at some point, or updating from OS X 10.10 to 10.11 somehow broke this. I'm using a custom font (a patched version of Menlo for Powerline) so I know the font didn't change out from under me.

To see this, try pasting ⟶ or ⟹ into MacVim with the `'guifont'` set to `Menlo:h11` and "Use Core Text renderer" enabled. Here's a screenshot showing it occurring with `mvim -N -u NONE`:

<img width="304" alt="screen shot 2016-02-14 at 4 15 29 pm" src="https://cloud.githubusercontent.com/assets/714/13037473/3917f99c-d336-11e5-8e97-70f1fa8ba6b1.png">

Note: When the CoreText renderer is disabled, MacVim does use font fallback, but it doesn't handle line height properly, which I'll file as a separate bug.

OS X 10.11.3 (15D21)

```
> vim --version
VIM - Vi IMproved 7.4 (2013 Aug 10, compiled Feb 14 2016 15:39:36)
MacOS X (unix) version
Included patches: 1-1265
Compiled by Homebrew
Huge version with MacVim GUI.  Features included (+) or not (-):
+acl             +farsi           +mouse_netterm   +syntax
+arabic          +file_in_path    +mouse_sgr       +tag_binary
+autocmd         +find_in_path    -mouse_sysmouse  +tag_old_static
+balloon_eval    +float           +mouse_urxvt     -tag_any_white
+browse          +folding         +mouse_xterm     -tcl
++builtin_terms  -footer          +multi_byte      +terminfo
+byte_offset     +fork()          +multi_lang      +termresponse
+channel         +fullscreen      -mzscheme        +textobjects
+cindent         -gettext         +netbeans_intg   +title
+clientserver    -hangul_input    +odbeditor       +toolbar
+clipboard       +iconv           +path_extra      +transparency
+cmdline_compl   +insert_expand   +perl            +user_commands
+cmdline_hist    +jumplist        +persistent_undo +vertsplit
+cmdline_info    +keymap          +postscript      +virtualedit
+comments        +langmap         +printer         +visual
+conceal         +libcall         +profile         +visualextra
+cryptv          +linebreak       +python          +viminfo
+cscope          +lispindent      -python3         +vreplace
+cursorbind      +listcmds        +quickfix        +wildignore
+cursorshape     +localmap        +reltime         +wildmenu
+dialog_con_gui  +lua             +rightleft       +windows
+diff            +menu            +ruby            +writebackup
+digraphs        +mksession       +scrollbind      -X11
+dnd             +modify_fname    +signs           -xfontset
-ebcdic          +mouse           +smartindent     +xim
+emacs_tags      +mouseshape      -sniff           -xsmp
+eval            +mouse_dec       +startuptime     -xterm_clipboard
+ex_extra        -mouse_gpm       +statusline      -xterm_save
+extra_search    -mouse_jsbterm   -sun_workshop    -xpm
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "$VIM/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/Applications/MacVim.app/Contents/Resources/vim"
Compilation: clang -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_MACVIM -Wall -Wno-unknown-pragmas -pipe  -DMACOS_X_UNIX  -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -fno-strict-aliasing -fno-common -dynamic -arch x86_64 -arch i386 -g -Os -pipe -fno-common -fno-strict-aliasing -fwrapv -DENABLE_DTRACE -DMACOSX -DNDEBUG -Wall -Wstrict-prototypes -Wshorten-64-to-32 -DNDEBUG -g -fwrapv -Os -Wall -Wstrict-prototypes -DENABLE_DTRACE -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.11 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1      
Linking: clang   -L. -L/usr/local/lib -L. -L/usr/local/lib -L/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/config -lpython2.7 -framework CoreFoundation -Wl,-syslibroot,/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -mmacosx-version-min=10.11 -L/usr/local/lib -o Vim -framework Cocoa -framework Carbon       -lm  -lncurses -liconv -framework Cocoa  -L/usr/local/lib -llua -fstack-protector  -L/System/Library/Perl/5.18/darwin-thread-multi-2level/CORE -lperl -framework Python   -framework Ruby    
```

**Comment by splhack** [comment]
> But I'm not quite sure when it broke. Either MacVim used to do font fallback and stopped at some point, or updating from OS X 10.10 to 10.11 somehow broke this.

@kballard Could you specify the latest working version of MacVim? Download binary from https://github.com/macvim-dev/macvim/releases

**Comment by kballard** [comment]
What I'm using is the latest version that Homebrew has, which is version `7.4-96`. Isn't this the same thing as the snapshot-96 listed on that releases page? It's the same patchlevel.

**Comment by kballard** [comment]
The error reproduces with the binary from the snapshot-96 DMG.

**Comment by splhack** [comment]
@kballard Could you specify the latest **working** version of MacVim? Some version works you said.

**Comment by kballard** [comment]
Oh I see what you mean. As I said in the ticket, I don't know if this broke with a MacVim update or an OS update. It's been over a year since I last tried to edit files with these particular glyphs. I'll try downloading an older build and seeing if that does anything.

**Comment by kballard** [comment]
Ah hah! Snapshot 74 does indeed have font fallback! I'll go through the other snapshots now to figure out when it broke.

**Comment by kballard** [comment]
@splhack It works in snapshot-88 (7.4.979) but it's broken in snapshot-89 (7.4.1054).

**Comment by splhack** [comment]
@kballard good catch. Could you do `git bisect`?

**Comment by kballard** [comment]
I'll try.

**Comment by splhack** [comment]
- Download https://raw.githubusercontent.com/macvim-dev/homebrew-macvim/master/macvim.rb
- Start git bisect

    $ git bisect start
    $ git bisect good snapshot-88
    $ git bisect bad snapshot-89

- Edit macvim.rb

    $ git status
    HEAD detached at 92c23d8

    head 'https://github.com/macvim-dev/macvim.git', :revision => "92c23d8"

- Install MacVim

    $ brew install --HEAD macvim.rb

- Test

set good or bad

    $ git bisect good

or

    $ git bisect bad

**Comment by kballard** [comment]
I'm just compiling from source. BTW, you can't just do a straight bisect like that, because it'll land on commits from the upstream vim instead of just limiting itself to the first parent, which means most tested commits won't be capable of building MacVim.app.

**Comment by splhack** [comment]
`git bisect skip` in that case.

**Comment by kballard** [comment]
Yeah I had to throw in a `git bisect skip snapshot-88.master^2` to tell it to skip everything that was merged in from upstream (since `master^2` is the latest upstream point that was merged in).

**Comment by kballard** [comment]
`git bisect` says the culprit is:

```
commit 3def55116e4ea06503c3a3496ffb8492e1a67738
Author: Kazuki Sakamoto <sakamoto@splhack.org>
Date:   Fri Dec 18 23:04:51 2015 -0800

    Remove legacy pre-10.7 stuff
    
    Fix #173
```

**Comment by kballard** [comment]
Ugh, belay that. With the parent commit, it looks fine, until I turn the CoreText renderer off and back on again, at which point it breaks. I'm guessing it wasn't actually using the CoreText renderer because the removal of ATSUI changed the numbering.

**Comment by kballard** [comment]
Yeah, snapshot 88 is broken too. Ugh. Time to check the older snapshots again.

I don't know why you didn't just keep the renderer preference numbering the same when taking out ATSUI :/

**Comment by kballard** [comment]
Yeah, even snapshot 73 is now broken. I can't run any previously-built snapshots because they won't launch on Yosemite (tries to load a libperl.dylib that doesn't exist). I can build snapshot 72 myself, which continues to reproduce the error, but I can't even build any older snapshots because of a compile error in os_unix.c (redeclaration of signalstack).

I'm wondering at this point if there's any chance that what's going on here is a linked-on-or-after check. It's possible that MacVim, when built on an older OS, will have font fallback behavior, but when built on Yosemite will not. I don't know if this is true, but it would explain why this behavior used to work, and why I can't find a version that I can run that shows the working behavior now.

**Comment by kballard** [comment]
After looking at how MMCoreTextView is actually implemented, it does appear to actually be doing font fallback manually. It's using `CTFontGetGlyphsForCharacters()` and `CTFontDrawGlyphs()` for the actual drawing, instead of using any of the higher-level CoreText features, so that doesn't handle fallback. But the code here explicitly checks for glyphs that can't be found in the given font and tries to find a fallback.

Which is to say, I'm not sure why it's failing. `CTFontCreateForString()` says it tries to find a font from the cascade list of a given font, so maybe the given font has no cascade list and maybe CoreText isn't falling back to looking at other fonts? I'm really not sure.

**Comment by kballard** [comment]
Yeah, the font fallback is ending up with the .LastResort font.

**Comment by kballard** [comment]
I think what happened here is the global cascade list changed between OS releases. Although I'm pretty surprised Apple Symbols isn't in the cascade list at all.

**Comment by kballard** [comment]
PR submitted as #245.

**Comment by splhack** [comment]
> a patched version of Menlo for Powerline

Why don't you share the url? This one? https://github.com/abertsch/Menlo-for-Powerline/blob/master/Menlo%20for%20Powerline.ttf

**Comment by kballard** [comment]
No, one I patched myself using a modified version of the script in the Powerline repo. But the problem reproduces with the built-in Menlo too.

**Comment by splhack** [comment]
Anyway I can't reproduce this issue. How to set up the system' Menlo font and what is the exact text to test?

    $ defaults delete org.vim.MacVim
    $ /Applications/MacVim.app/Contents/MacOS/Vim -g -u NONE -U NONE

**Comment by kballard** [comment]
Huh, I just tried on a different computer and the problem isn't occurring here. That's pretty strange. I just verified that `CTFontCreateForString()`, when given the Menlo font and the string `"⟶"`, still returns the `.LastResort` font, so I have no idea why MacVim is managing to still draw the correct glyph on this computer. My only guess is that somehow the `fontCache` contains one of the fonts that can draw the glyph.

**Comment by kballard** [comment]
Nope, it's definitely getting the font back from `CTFontCreateForString()`. But when I try calling that in a sample project, I'm getting `.LastResort` back. This is rather mystifying.

However, this did reveal another issue: once MacVim has gotten the `.LastResort` font back from something, that font ends up in its font cache, which means it'll pull glyphs from `.LastResort` instead of trying to find a new font.

**Comment by kballard** [comment]
This also makes me question the current behavior of trying to find one font for the whole string, then halving repeatedly. If it comes up with a sub-par font for the whole range, it'll end up using that font even if it would do better to use multiple different fonts.

I'm wondering if maybe my PR #245, which uses `CTLine` to assign fonts, should actually be refactored further to get rid of the `lookupFont()` function entirely, and the recursive nature of the string drawing, and instead just create a `CTLine` for the whole string at once. Then it can just iterate over the runs, grabbing glyphs and fonts and drawing them itself. No more recursion, no more searching for fonts, and no more `fontCache` either (since there's no way to expose that to `CTLine`, and presumably CoreText has its own cache anyway).

**Comment by splhack** [comment]
I'm not sure I understand this issue correctly, but we can close this issue because no one can reproduce this issue except some ill-formed environment, right?
And I wasn't convinced enough that #245 can solve something or improve something. Could you provide more information? like, the amount of memory usage, CPU usage, Instruments profiling result, ...

**Comment by kballard** [comment]
My environment isn't ill-formed. I don't know why I'm seeing it on one computer and not another, but my computer that does reproduce this is a perfectly legitimate environment. Especially because I'm not actually sure how it even works correctly on my other computer, since calling `CTFontCreateForString()` in a sample project given the font and string that MacVim are using always comes up with `.LastResort`.

**Comment by kballard** [comment]
If you can answer this question, that would help: Why does `CTFontCreateForString()`, given Menlo as the font and `"⟶"` as the string, return STIXGeneral in MacVim (on my laptop), but running the equivalent code in a standalone project returns .LastResort (which is what happens on my desktop)? Or rather, I get why it returns .LastResort, since STIXGeneral isn't in the default cascade list and the Menlo `CTFontRef` doesn't contain a custom cascade list, so the real question is why is it returning STIXGeneral at all?

**Comment by CarterLi** [comment]
I think I encountered the same issue. Here are snapshots (~/.vimrc):
*  Without `Use Core Text renderer` enabled. Incorrect line height, see the airline status bar:
![image](https://cloud.githubusercontent.com/assets/6134068/15400691/fd733e68-1e1e-11e6-89cc-a126632ab6dc.png)
* With `Use Core Text renderer` enabled, lots of `?`, see the right side of the airline status bar:
![image](https://cloud.githubusercontent.com/assets/6134068/15400731/232858a0-1e1f-11e6-850f-dac54ae1de97.png)

**Comment by symbolix** [comment]
Same here. I am more than happy to provide a screen-shot but not different than anything we already have here. An annoying [?] character in the status bar when running MacVim and no issues when running `/Applications/MacVim.app/Contents/MacOS/Vim` in iTerm2. I have compiled that latest snapshot   `105` of MacVim from source and still the issue exists.

**Comment by CarterLi** [comment]
Have no idea what's happening, but upgrading to 10.12 solved this issue...

**Comment by dashingspidy** [comment]
I'm using Macvim 8.0 build 110  on 10.12
If i uncheck `Use core text renderer`, macligatures take effects immediately (while core text checked it take effects only after save the file) & Vim-airline works fine. 

But linespace isn't working properly. I was trying to get macligatures on. Found this unexpected solution. 

I don't know what happened but i guess it's an unexpected behaviour.

**Comment by exchgr** [comment]
Confirmed on 8.0-114 through 127, macOS Sierra 10.12.3. This screenshot shows special characters from NERDtree and neomake:
![screen shot 2017-03-20 at 3 29 08 pm](https://cloud.githubusercontent.com/assets/824678/24117839/16f4dd76-0d82-11e7-803d-1a7bf14a9536.png)

**Comment by exchgr** [comment]
(Here it is without Core Text)
![screen shot 2017-03-20 at 3 35 35 pm](https://cloud.githubusercontent.com/assets/824678/24118077/ed444d44-0d82-11e7-9c9e-b9436a73e8af.png)
